File: playground/negotiation-ecs/SDKs/go/agent_client.go
// Command agent_client is a Go SDK example for the Multi-Agent Negotiation
// Simulator. It connects over gRPC, streams simulation frames, and participates
// in negotiations as an external agent.
//
// Protocol: an external agent controls one body. Its first proposal claims a
// body (the engine returns the assigned entity id); subsequent proposals are
// decisions for that body when it is the agent's turn to respond. A decision is
// expressed as the offer payload's "action" field — "accept", "reject", or
// (default) a counter-offer.
//
// Usage:
//
//	go run . [--server localhost:50051] [--agent-id go-coop-001]
package main

import (
	"context"
	"encoding/json"
	"flag"
	"io"
	"log"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	pb "github.com/erancihan/negotiation-ecs/backend-go/gen/proto/negotiationpb"
)

// Agent is the interface an external agent implements.
type Agent interface {
	// OnFrame observes the world each tick.
	OnFrame(frame *pb.SimFrame)
	// Decide returns a proposal to submit, or nil to do nothing this frame.
	Decide(frame *pb.SimFrame) *pb.Proposal
	// OnAck receives the result of a submitted proposal (carries the assigned
	// body id on first contact).
	OnAck(ack *pb.ProposalAck)
}

// CooperativeTrader claims a body, then accepts whenever it is its turn to
// respond. A complete, protocol-correct example.
type CooperativeTrader struct {
	AgentID  string
	assigned uint64 // wire id of the body we control (0 until assigned)
	claimed  bool
}

func (a *CooperativeTrader) OnFrame(frame *pb.SimFrame) {
	if frame.GetTick()%100 == 0 {
		log.Printf("[%s] tick %d | body=%d | entities=%d",
			a.AgentID, frame.GetTick(), a.assigned, len(frame.GetEntities()))
	}
}

func (a *CooperativeTrader) Decide(frame *pb.SimFrame) *pb.Proposal {
	// Claim a body on first contact.
	if !a.claimed {
		a.claimed = true
		return a.proposal(map[string]any{"action": "counter", "hello": true})
	}
	if a.assigned == 0 {
		return nil // waiting for the assignment ack
	}
	// If our body is responding (COUNTERING), accept the standing offer.
	for _, e := range frame.GetEntities() {
		if e.GetEntityId() == a.assigned &&
			e.GetNegotiationStatus() == pb.NegotiationStatus_NEGOTIATION_STATUS_COUNTERING {
			return a.proposal(map[string]any{"action": "accept"})
		}
	}
	return nil
}

func (a *CooperativeTrader) OnAck(ack *pb.ProposalAck) {
	if a.assigned == 0 && ack.GetAccepted() {
		a.assigned = ack.GetAssignedEntityId()
		log.Printf("[%s] assigned body %d", a.AgentID, a.assigned)
	}
}

func (a *CooperativeTrader) proposal(offer map[string]any) *pb.Proposal {
	js, _ := json.Marshal(offer)
	return &pb.Proposal{AgentId: a.AgentID, OfferJson: string(js)}
}

func main() {
	serverAddr := flag.String("server", "localhost:50051", "gRPC server address")
	agentID := flag.String("agent-id", "go-coop-001", "agent identifier")
	flag.Parse()

	conn, err := grpc.NewClient(*serverAddr, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		log.Fatalf("connect: %v", err)
	}
	defer conn.Close()

	client := pb.NewSimulationServiceClient(conn)
	agent := &CooperativeTrader{AgentID: *agentID}
	log.Printf("[%s] connected to %s", *agentID, *serverAddr)

	stream, err := client.StreamSimulation(context.Background(), &pb.StreamRequest{MaxFps: 10})
	if err != nil {
		log.Fatalf("stream: %v", err)
	}

	for {
		frame, err := stream.Recv()
		if err == io.EOF {
			log.Printf("[%s] stream ended", *agentID)
			return
		}
		if err != nil {
			log.Fatalf("[%s] stream error: %v", *agentID, err)
		}

		agent.OnFrame(frame)
		if proposal := agent.Decide(frame); proposal != nil {
			ack, err := client.SubmitProposal(context.Background(), proposal)
			if err != nil {
				log.Printf("[%s] proposal error: %v", *agentID, err)
				continue
			}
			agent.OnAck(ack)
		}
	}
}

File: playground/negotiation-ecs/SDKs/python/agent_client.py
"""
Multi-Agent Negotiation Simulator — Python Agent SDK

An external agent controls one body. Its first proposal claims a body (the engine
returns the assigned entity id in the ack); subsequent proposals are decisions for
that body when it is the agent's turn to respond. A decision is expressed by the
offer payload's "action" field — "accept", "reject", or (default) a counter-offer.

Setup (one-time): generate the gRPC stubs from the project root:
    make sdk-python
    # or: cd SDKs/python && python -m grpc_tools.protoc \
    #       -I../../proto --python_out=. --grpc_python_out=. ../../proto/simulation.proto

Usage:
    python agent_client.py [--server localhost:50051] [--agent-id py-coop-001]
"""

import argparse
import json
from abc import ABC, abstractmethod

import grpc

import simulation_pb2
import simulation_pb2_grpc


class NegotiationAgent(ABC):
    """Base class for external negotiation agents.

    Subclass and implement ``on_frame`` and ``decide``. The agent loop streams
    frames, submits the decisions ``decide`` returns, and routes acks back via
    ``on_proposal_result`` (where the assigned body id arrives on first contact).
    """

    def __init__(self, agent_id: str, server_addr: str = "localhost:50051"):
        self.agent_id = agent_id
        self.server_addr = server_addr
        self.assigned_entity_id = 0
        self._channel = None
        self._stub = None

    def run(self):
        self._channel = grpc.insecure_channel(self.server_addr)
        self._stub = simulation_pb2_grpc.SimulationServiceStub(self._channel)
        print(f"[{self.agent_id}] connected to {self.server_addr}")
        try:
            stream = self._stub.StreamSimulation(simulation_pb2.StreamRequest(max_fps=10))
            for frame in stream:
                self.on_frame(frame)
                offer = self.decide(frame)
                if offer is not None:
                    ack = self._stub.SubmitProposal(simulation_pb2.Proposal(
                        agent_id=self.agent_id,
                        offer_json=json.dumps(offer),
                    ))
                    self.on_proposal_result(ack)
        except grpc.RpcError as e:
            print(f"[{self.agent_id}] stream error: {e}")
        except KeyboardInterrupt:
            print(f"\n[{self.agent_id}] interrupted")
        finally:
            if self._channel:
                self._channel.close()

    def on_proposal_result(self, ack):
        if self.assigned_entity_id == 0 and ack.accepted:
            self.assigned_entity_id = ack.assigned_entity_id
            print(f"[{self.agent_id}] assigned body {self.assigned_entity_id}")

    @abstractmethod
    def on_frame(self, frame):
        """Observe the world each tick."""

    @abstractmethod
    def decide(self, frame):
        """Return an offer dict to submit, or None to do nothing this frame."""


class CooperativeTrader(NegotiationAgent):
    """Claims a body, then accepts whenever it is its turn to respond."""

    def __init__(self, agent_id: str, server_addr: str = "localhost:50051"):
        super().__init__(agent_id, server_addr)
        self._claimed = False

    def on_frame(self, frame):
        if frame.tick % 100 == 0:
            print(f"[{self.agent_id}] tick {frame.tick} | body={self.assigned_entity_id} "
                  f"| entities={len(frame.entities)}")

    def decide(self, frame):
        # Claim a body on first contact.
        if not self._claimed:
            self._claimed = True
            return {"action": "counter", "hello": True}
        if self.assigned_entity_id == 0:
            return None  # waiting for the assignment ack
        # If our body is responding (COUNTERING), accept the standing offer.
        for e in frame.entities:
            if (e.entity_id == self.assigned_entity_id
                    and e.negotiation_status == simulation_pb2.NEGOTIATION_STATUS_COUNTERING):
                return {"action": "accept"}
        return None


def main():
    parser = argparse.ArgumentParser(description="Negotiation Agent SDK - Python")
    parser.add_argument("--server", default="localhost:50051", help="gRPC server address")
    parser.add_argument("--agent-id", default="py-coop-001", help="agent identifier")
    args = parser.parse_args()

    CooperativeTrader(agent_id=args.agent_id, server_addr=args.server).run()


if __name__ == "__main__":
    main()
